File: src/minima_harness/ai/providers/faux.py
"""In-memory provider for hermetic tests and demos.

Minimal port of PI's ``registerFauxProvider``. Opt-in; not registered by default.
One deterministic scripted flow per registration.
Usage is estimated at roughly 1 token per 4 characters when not provided on the message.
"""

from __future__ import annotations

from collections import deque
from collections.abc import AsyncIterator
from typing import TYPE_CHECKING

from minima_harness.ai.events import (
    DoneEvent,
    ErrorEvent,
    StartEvent,
    TextDeltaEvent,
    TextEndEvent,
    TextStartEvent,
    ThinkingDeltaEvent,
    ThinkingEndEvent,
    ThinkingStartEvent,
    ToolCallEndEvent,
    ToolCallStartEvent,
)
from minima_harness.ai.providers.base import Provider, register_provider, unregister_provider
from minima_harness.ai.types import (
    AssistantMessage,
    Modality,
    Model,
    ModelCost,
    TextContent,
    ThinkingContent,
    ToolCall,
)
from minima_harness.ai.usage import attach_cost

if TYPE_CHECKING:
    from minima_harness.ai.events import Event


_FAUX_MODEL = Model(
    id="faux",
    provider="faux",
    api="faux",
    name="Faux (test)",
    cost=ModelCost(input=0.0, output=0.0),
    context_window=8192,
    max_tokens=4096,
    input=(Modality.text,),
    reasoning=False,
)

# Roughly 1 token per 4 characters, per PI's faux provider.
_CHARS_PER_TOKEN = 4


def _estimate_usage(msg: AssistantMessage) -> None:
    if msg.usage.input or msg.usage.output:
        return
    char_len = sum(
        len(b.text) if isinstance(b, TextContent) else len(getattr(b, "thinking", ""))
        for b in msg.content
        if not isinstance(b, str)
    )
    msg.usage.output = max(1, char_len // _CHARS_PER_TOKEN)


class FauxProviderState:
    """Observable per-registration state."""

    def __init__(self) -> None:
        self.call_count = 0
        self.responses: deque[AssistantMessage] = deque()

    @property
    def pending_response_count(self) -> int:
        return len(self.responses)


class FauxRegistration:
    """Handle returned by :func:`register_faux_provider`."""

    def __init__(self, *, models: list[Model] | None = None) -> None:
        self.models = models or [_FAUX_MODEL]
        self.state = FauxProviderState()
        self._provider = _FauxProvider(self.state, self.models)

    def get_model(self, model_id: str | None = None) -> Model:
        if model_id is None:
            return self.models[0]
        for m in self.models:
            if m.id == model_id:
                return m
        raise KeyError(model_id)

    def set_responses(self, messages: list[AssistantMessage]) -> None:
        self.state.responses = deque(messages)

    def append_responses(self, messages: list[AssistantMessage]) -> None:
        self.state.responses.extend(messages)

    def register(self) -> FauxRegistration:
        register_provider("faux", self._provider)
        return self

    def unregister(self) -> None:
        unregister_provider("faux")

    def __enter__(self) -> FauxRegistration:
        register_provider("faux", self._provider)
        return self

    def __exit__(self, *exc: object) -> None:
        unregister_provider("faux")


class _FauxProvider(Provider):
    def __init__(self, state: FauxProviderState, models: list[Model]) -> None:
        self.api_id = "faux"
        self.state = state
        self.models = models

    async def stream(
        self,
        model: Model,
        context: object,
        *,
        options: dict | None = None,
        signal: object | None = None,
    ) -> AsyncIterator[Event]:
        self.state.call_count += 1
        if not self.state.responses:
            err = AssistantMessage(
                content=[TextContent(text="")],
                stop_reason="error",
                error_message="No more faux responses queued",
            )
            err.model = model.id
            yield ErrorEvent(reason="error", error=err)
            return

        msg = self.state.responses.popleft()
        msg.model = model.id
        _estimate_usage(msg)
        attach_cost(model, msg.usage)

        yield StartEvent(partial=msg)
        for index, block in enumerate(msg.content):
            if isinstance(block, TextContent):
                yield TextStartEvent(content_index=index)
                if block.text:
                    yield TextDeltaEvent(delta=block.text, content_index=index)
                yield TextEndEvent(content=block.text, content_index=index)
            elif isinstance(block, ThinkingContent):
                yield ThinkingStartEvent(content_index=index)
                if block.thinking:
                    yield ThinkingDeltaEvent(delta=block.thinking, content_index=index)
                yield ThinkingEndEvent(content=block.thinking, content_index=index)
            elif isinstance(block, ToolCall):
                yield ToolCallStartEvent(content_index=index)
                yield ToolCallEndEvent(tool_call=block, content_index=index)
        yield DoneEvent(reason=msg.stop_reason, message=msg)


def register_faux_provider(*, models: list[Model] | None = None) -> FauxRegistration:
    """Register a temporary in-memory provider for tests/demos.

    Remember to call ``.unregister()`` (or use the provider as a context manager) so the
    faux api id does not leak across tests.
    """
    return FauxRegistration(models=models).register()

File: packages/tui/src/ai/providers/faux.ts
/**
 * In-memory provider for hermetic tests and demos.
 *
 * Port of minima_harness/ai/providers/faux.py. Opt-in; not registered by default.
 * One deterministic scripted flow per registration. Usage is estimated at roughly
 * 1 token per 4 characters when not provided on the message.
 */

import type { StreamEvent } from "../events.ts";
import {
  done as doneEv,
  error as errorEv,
  start as startEv,
  textDelta,
  textEnd,
  textStart,
  thinkingDelta,
  thinkingEnd,
  thinkingStart,
  toolCallEnd,
  toolCallStart,
} from "../events.ts";
import { AssistantMessage, type Context, type Model, text } from "../types.ts";
import { attachCost } from "../usage.ts";
import { type Provider, registerProvider, unregisterProvider } from "./base.ts";

const FAUX_MODEL: Model = {
  id: "faux",
  provider: "faux",
  api: "faux",
  name: "Faux (test)",
  cost: { input: 0, output: 0 },
  context_window: 8192,
  max_tokens: 4096,
  input: ["text"],
  reasoning: false,
};

// Roughly 1 token per 4 characters, per PI's faux provider.
const CHARS_PER_TOKEN = 4;

function estimateUsage(msg: AssistantMessage): void {
  if (msg.usage.input || msg.usage.output) return;
  let charLen = 0;
  for (const b of msg.content) {
    if (b.type === "text") charLen += b.text.length;
    else if (b.type === "thinking") charLen += b.thinking.length;
  }
  msg.usage.output = Math.max(1, Math.floor(charLen / CHARS_PER_TOKEN));
}

/** Observable per-registration state. */
export class FauxProviderState {
  callCount = 0;
  responses: AssistantMessage[] = [];
  get pendingResponseCount(): number {
    return this.responses.length;
  }
}

/** Handle returned by registerFauxProvider. */
export class FauxRegistration {
  readonly models: Model[];
  readonly state = new FauxProviderState();
  private readonly provider: FauxProvider;

  constructor(models?: Model[]) {
    this.models = models ?? [FAUX_MODEL];
    this.provider = new FauxProvider(this.state, this.models);
  }

  getModel(modelId?: string): Model {
    if (!modelId) return this.models[0]!;
    const m = this.models.find((x) => x.id === modelId);
    if (!m) throw new Error(`faux model not found: ${modelId}`);
    return m;
  }

  setResponses(messages: AssistantMessage[]): this {
    this.state.responses = [...messages];
    return this;
  }

  appendResponses(messages: AssistantMessage[]): this {
    this.state.responses.push(...messages);
    return this;
  }

  register(): this {
    registerProvider("faux", this.provider);
    return this;
  }

  unregister(): void {
    unregisterProvider("faux");
  }

  /** Use as a callback: registers, runs `fn`, always unregisters. */
  async use<T>(fn: () => Promise<T>): Promise<T> {
    this.register();
    try {
      return await fn();
    } finally {
      this.unregister();
    }
  }
}

class FauxProvider implements Provider {
  readonly apiId = "faux";
  constructor(
    private readonly state: FauxProviderState,
    private readonly models: Model[],
  ) {}

  async *stream(
    model: Model,
    _context: Context,
    opts?: { options?: Record<string, unknown>; signal?: AbortSignal },
  ): AsyncIterable<StreamEvent> {
    // Honor a pre-aborted signal like the real providers, so abort behaviour is
    // exercisable in hermetic tests.
    if (opts?.signal?.aborted) throw new DOMException("Aborted", "AbortError");
    this.state.callCount += 1;
    const queued = this.state.responses.shift();
    if (!queued) {
      const err = new AssistantMessage({
        content: [text("")],
        stop_reason: "error",
        error_message: "No more faux responses queued",
      });
      err.model = model.id;
      yield errorEv("error", err);
      return;
    }

    const msg = queued;
    msg.model = model.id;
    estimateUsage(msg);
    attachCost(model, msg.usage);

    yield startEv(msg);
    for (const [index, block] of msg.content.entries()) {
      if (block.type === "text") {
        yield textStart(index);
        if (block.text) yield textDelta(block.text, index);
        yield textEnd(block.text, index);
      } else if (block.type === "thinking") {
        yield thinkingStart(index);
        if (block.thinking) yield thinkingDelta(block.thinking, index);
        yield thinkingEnd(block.thinking, index);
      } else if (block.type === "toolCall") {
        yield toolCallStart(index);
        yield toolCallEnd(block, index);
      }
    }
    yield doneEv(msg.stop_reason, msg);
  }
}

/** Register a temporary in-memory provider for tests/demos. Unregister when done. */
export function registerFauxProvider(models?: Model[]): FauxRegistration {
  return new FauxRegistration(models).register();
}
